envoy-control-tests/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/config/EnvoyControlTestApp.kt

```diff
@@ -83,15 +85,21 @@ class EnvoyControlRunnerTestApp(
         return objectMapper.readValue(response.body()?.use { it.string() }, ServicesState::class.java)
     }
 
-    override fun getSnapshot(nodeJson: String): String {
+    override fun getSnapshot(nodeJson: String): SnapshotDebugResponse {
         val response = httpClient.newCall(
             Request.Builder()
                 .post(RequestBody.create(MediaType.get("application/json"), nodeJson))
                 .url("http://localhost:$appPort/snapshot")
                 .build()
         ).execute()
 
-        return response.body().use { it!!.string() }
+        if (!response.isSuccessful) {
+            return SnapshotDebugResponse(found = false)
+        }
+
+        return response.body()
+            .use { objectMapper.readValue(it!!.byteStream(), SnapshotDebugResponse::class.java) }
```

Comment:
I would make this error more explicit 
```suggestion
            ?.use { objectMapper.readValue(it.byteStream(), SnapshotDebugResponse::class.java) }
            ?.copy(found = true) ?: throw SnapshotDebugResponseMissingException()
    }

    class SnapshotDebugResponseMissingException :
        RuntimeException("Expected snapshot debug in response body but got none")
```